### PR TITLE
Alpha support for provision volumes from cross-namespace data sources

### PR DESCRIPTION
**Pull request by ttakahashi21**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Please see [CrossNamespaceVolumeDataSource](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots)

**Which issue(s) this PR fixes**:

Part of #https://github.com/kubernetes-csi/external-provisioner/issues/804

**Special notes for your reviewer**:
Discuss the API definition https://github.com/kubernetes/kubernetes/pull/113186.
The code may change due to API specification changes.

**Does this PR introduce a user-facing change?**:

```release-note
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots
```

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -57,6 +59,7 @@ import (
 	"github.com/kubernetes-csi/csi-lib-utils/rpc"
 	snapapi "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
 	snapclientset "github.com/kubernetes-csi/external-snapshotter/client/v6/clientset/versioned"
+	referenceGrantv1alpha2 "sigs.k8s.io/gateway-api/pkg/client/listers/apis/v1alpha2"
```
Comment (humblec) [inline pkg/controller/controller.go:62]: just wondering are we fine to use alphav2 of RefGrants? considering we have v1beta1 avaiable , cant we or better to use it ?

File: pkg/features/features.go [hunk 2]
```diff
@@ -43,6 +50,7 @@ func init() {
 // defaultKubernetesFeatureGates consists of all known feature keys specific to external-provisioner.
 // To add a new feature, define a key for it above and add it here.
 var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{
-	Topology:             {Default: false, PreRelease: featuregate.GA},
-	HonorPVReclaimPolicy: {Default: false, PreRelease: featuregate.Alpha},
+	Topology:                         {Default: false, PreRelease: featuregate.GA},
+	HonorPVReclaimPolicy:             {Default: false, PreRelease: featuregate.Alpha},
+	CrossNamespaceSourceProvisioning: {Default: false, PreRelease: featuregate.Alpha},
```
Comment (humblec) [inline pkg/features/features.go:55]: we also need to add this feature gate to the readme doc

**Comment by humblec** [comment]
@ttakahashi21 thanks.  I have few review comments . Apart from that, can we add some tests to this PR ?

File: pkg/controller/controller.go (hunk 1 above)
Comment (xing-yang) [reply pkg/controller/controller.go:62]: +1

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -1081,47 +1126,114 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
...
+}
+
+// getSnapshotSourceDataSourceRef2 verifies DataSourceRef2.Kind of type VolumeSnapshot, making sure that the requested Snapshot is available/ready
+// returns the VolumeContentSource for the requested snapshot
+func (p *csiProvisioner) getSnapshotSourceDataSourceRef2(ctx context.Context, claim *v1.PersistentVolumeClaim, sc *storagev1.StorageClass) (*csi.VolumeContentSource, error) {
+	if claim.Spec.DataSourceRef2.Namespace == "" {
+		return p.getSnapshotSourceInternal(ctx, claim, sc, claim.Namespace, claim.Spec.DataSourceRef2.Name)
+	}
+
+	if ok, err := p.IsGranted(ctx, claim.Spec.DataSourceRef2.Namespace, claim.Spec.DataSourceRef2.Name, claim); err != nil || !ok {
```
Comment (xing-yang) [inline pkg/controller/controller.go:1190]: Do we check somewhere if the PVC namespace is the same as the VolumeSnapshot namespace?  If they are the same, then referenceGrant is not needed.

File: pkg/controller/controller.go (hunk 1 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go:62]: The ReferenceGrant is not yet in beta. Therefore, I'm using Alpha2.
See the URL below.
https://github.com/kubernetes-sigs/gateway-api

File: pkg/features/features.go (hunk 2 above)
Comment (ttakahashi21) [reply pkg/features/features.go]: Could you tell me where should i write it?

File: pkg/controller/controller.go (hunk 1 above)
Comment (humblec) [reply pkg/controller/controller.go:62]: @ttakahashi21 isnt this https://github.com/kubernetes-sigs/gateway-api/tree/main/pkg/client/listers/apis/v1beta1  available ?

Comment (ttakahashi21) [reply pkg/controller/controller.go:62]: Thank you for letting me know.

Comment (ttakahashi21) [reply pkg/controller/controller.go]: I was able to use it by adding the following to my go.mod file:
> require sigs.k8s.io/gateway-api v0.5.1-0.20221110191515-2d586cc49ab4

File: pkg/features/features.go (hunk 2 above)
Comment (ttakahashi21) [reply pkg/features/features.go]: Added CrossNamespaceVolumeDataSource in [README.md](https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5).

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -582,6 +589,21 @@ func (p *csiProvisioner) prepareProvision(ctx context.Context, claim *v1.Persist
...
+	if utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) {
+		if claim.Spec.DataSourceRef != nil && claim.Spec.DataSourceRef.Namespace != nil && len(*claim.Spec.DataSourceRef.Namespace) > 0 {
+			switch claim.Spec.DataSourceRef.Kind {
+			case snapshotKind:
+				if *(claim.Spec.DataSourceRef.APIGroup) != snapshotAPIGroup {
+					return nil, controller.ProvisioningFinished, fmt.Errorf("the PVC source does not belong to the right APIGroup. Expected %s, Got %s", snapshotAPIGroup, *(claim.Spec.DataSourceRef.APIGroup))
+				}
+				rc.snapshot = true
+			default:
+				return nil, controller.ProvisioningFinished, fmt.Errorf("the PVC source specified via DataSourceRef with non-empty namespace does not belong to the right Kind.  Expected %s, Got %s", snapshotKind, claim.Spec.DataSourceRef.Kind)
```
Comment (msau42) [inline pkg/controller/controller.go:602]: Are we allowing pvc source and data populators too?

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -1081,47 +1131,115 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
...
+				break
+			}
+		}
+		// Skip unrelated policy by checking From field
+		if !validFrom {
+			continue
+		}
+
+		for _, to := range grant.Spec.To {
+			if to.Group != snapshotAPIGroup || to.Kind != snapshotKind {
```
Comment (msau42) [inline pkg/controller/controller.go:1161]: Can we make the function more generic to handle non-snapshot data sources?

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -582,6 +589,21 @@ func (p *csiProvisioner) prepareProvision(ctx context.Context, claim *v1.Persist
 			}
 		}
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) {
+		if claim.Spec.DataSourceRef != nil && claim.Spec.DataSourceRef.Namespace != nil && len(*claim.Spec.DataSourceRef.Namespace) > 0 {
```
Comment (msau42) [inline pkg/controller/controller.go:594]: I think we should move the check for feature gate and return error earlier at this place. There's a number of csi calls we make after this that may be nice to avoid.

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -631,7 +653,17 @@ func (p *csiProvisioner) prepareProvision(ctx context.Context, claim *v1.Persist
 		},
 	}
 
-	if claim.Spec.DataSource != nil && (rc.clone || rc.snapshot) {
+	if claim.Spec.DataSourceRef != nil && claim.Spec.DataSourceRef.Namespace != nil && len(*claim.Spec.DataSourceRef.Namespace) > 0 {
+		if utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) && rc.snapshot {
+			volumeContentSource, err := p.getVolumeContentSourceFromXnsDataSource(ctx, claim, sc)
```
Comment (msau42) [inline pkg/controller/controller.go:658]: Do we need a new function, or can we modify `getVolumeContentSource`?

Comment (ttakahashi21) [reply pkg/controller/controller.go]: Yes. Is possible. Fixed code https://github.com/kubernetes-csi/external-provisioner/pull/805/commits/1f9df93e646bbd126bf7bcc172a7e2a3a2da4e8e
Could you check code?

File: pkg/controller/controller.go (hunk 6 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: I added feature gate check code early in the prepare Provision function.

File: pkg/controller/controller.go (hunk 4 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Yes. Added. https://github.com/kubernetes-csi/external-provisioner/commit/1f9df93e646bbd126bf7bcc172a7e2a3a2da4e8e

File: deploy/kubernetes/rbac.yaml [hunk 8]
```diff
@@ -57,6 +57,9 @@ rules:
   - apiGroups: ["storage.k8s.io"]
     resources: ["volumeattachments"]
     verbs: ["get", "list", "watch"]
+  - apiGroups: ["gateway.networking.k8s.io"]
```
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply deploy/kubernetes/rbac.yaml:60]: I see. But if I comment it out, CI fails. Is it possible to test without commenting out in the CI side test?

File: go.mod [hunk 9]
```diff
@@ -74,13 +76,13 @@ require (
...
 	k8s.io/kube-openapi v0.0.0-20220803164354-a70c9af30aea // indirect
-	k8s.io/utils v0.0.0-20220728103510-ee6ede2d64ed // indirect
+	k8s.io/utils v0.0.0-20221107191617-1a15be271d1d // indirect
 	sigs.k8s.io/json v0.0.0-20220713155537-f223a00ba0e2 // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
-replace k8s.io/api => k8s.io/api v0.25.2
+replace k8s.io/api => k8s.io/api v0.0.0-20221110160654-5cb32024090c
```
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply go.mod:86]: I set it temporarily because I need to use the API added for k8s 1.26 for successful CI testing.  Therefore, I will change it when the release tag is attached.

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -582,6 +594,28 @@ func (p *csiProvisioner) prepareProvision(ctx context.Context, claim *v1.Persist
 			}
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) {
+		if claim.Spec.DataSourceRef != nil {
```
Comment (msau42) [inline pkg/controller/controller.go:598]: This logic looks the same as the DataSource logic. Can we create a common function that takes in Name and Kind?

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -980,29 +1032,71 @@ func removePrefixedParameters(param map[string]string) (map[string]string, error
...
+		switch claim.Spec.DataSourceRef.Kind {
+		case snapshotKind:
+			return p.getSnapshotSource(ctx, claim, sc)
+		case pvcKind:
+			return p.getPVCSource(ctx, claim, sc)
+		default:
+			// For now we shouldn't pass other things to this function, but treat it as a noop and extend as needed
+			return nil, nil
+		}
+	} else {
```
Comment (msau42) [inline pkg/controller/controller.go:1045]: Do we need the if/else? The logic looks exactly the same.

File: pkg/controller/controller.go [hunk 12]
```diff
@@ -1081,47 +1175,117 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
 	return volumeContentSource, nil
 }
 
+func (p *csiProvisioner) IsGranted(ctx context.Context, claim *v1.PersistentVolumeClaim) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) {
+		return false, nil
+	}
+
+	// Get all ReferenceGrants in snapshot's namespace
```
Comment (msau42) [inline pkg/controller/controller.go:1183]: update comment to say "data source" instead of "snapshot"

File: pkg/controller/controller.go [hunk 13]
```diff
@@ -1081,47 +1175,117 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
 	return volumeContentSource, nil
 }
 
+func (p *csiProvisioner) IsGranted(ctx context.Context, claim *v1.PersistentVolumeClaim) (bool, error) {
```
Comment (msau42) [inline pkg/controller/controller.go:1178]: Add a todo that this function should eventually be replaced by a common kubernetes library

File: pkg/controller/controller.go [hunk 14]
```diff
@@ -1081,47 +1175,117 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
...
+				break
+			}
+		}
+		// Skip unrelated policy by checking From field
+		if !validFrom {
+			continue
+		}
+
+		for _, to := range grant.Spec.To {
+			if claim.Spec.DataSourceRef.APIGroup != nil && (string(to.Group) != *claim.Spec.DataSourceRef.APIGroup || string(to.Kind) != claim.Spec.DataSourceRef.Kind) {
```
Comment (msau42) [inline pkg/controller/controller.go:1205]: What if `dataSourceRef.APIGroup == nil` and `to.Group is set`?

Also it should be `groupMatches` and `kindMatches` right?

File: pkg/controller/controller.go [hunk 15]
```diff
@@ -1081,47 +1175,117 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
...
+			}
+		}
+
+		if allowed {
+			break
+		}
+	}
+
+	if !allowed {
+		return false, nil
```
Comment (msau42) [inline pkg/controller/controller.go:1220]: I think this could just be `return allowed, nil`

File: pkg/controller/controller.go [hunk 16]
```diff
@@ -924,7 +964,19 @@ func (p *csiProvisioner) Provision(ctx context.Context, options controller.Provi
 }
 
 func (p *csiProvisioner) setCloneFinalizer(ctx context.Context, pvc *v1.PersistentVolumeClaim) error {
-	claim, err := p.claimLister.PersistentVolumeClaims(pvc.Namespace).Get(pvc.Spec.DataSource.Name)
+	var claim *v1.PersistentVolumeClaim
+	var err error
+
+	if pvc.Spec.DataSourceRef != nil {
```
Comment (msau42) [inline pkg/controller/controller.go:970]: How much of this logic can be simplified across all the data sources if we had a common function that can normalize DataSource and DataSourceRef? Something like:
```
func dataSource(claim *v1.PersistentVolumeClaim) (ObjectReference, error)
```

I see that this pattern repeats many times for each of the data source types.

File: go.mod [hunk 17]
```diff
@@ -29,6 +29,8 @@ require (
 	sigs.k8s.io/sig-storage-lib-external-provisioner/v8 v8.0.0
 )
 
+require sigs.k8s.io/gateway-api v0.5.1-0.20221110191515-2d586cc49ab4
```
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply go.mod:32]: > Also is there a release tag of gateway-api we can use?
There is none.

@robscott
Could you please create a tag that includes the ReferenceGrant API Beta?

File: pkg/controller/controller.go [hunk 18]
```diff
@@ -537,6 +544,11 @@ func (p *csiProvisioner) prepareProvision(ctx context.Context, claim *v1.Persist
 		return nil, controller.ProvisioningFinished, errors.New("storage class was nil")
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) &&
+		claim.Spec.DataSourceRef != nil && claim.Spec.DataSourceRef.Namespace != nil && len(*claim.Spec.DataSourceRef.Namespace) > 0 {
+		return nil, controller.ProvisioningFinished, fmt.Errorf("error handling for DataSourceRef Type %s with non-empty namespace by Name %s: CrossNamespaceVolumeDataSource feature disabled", claim.Spec.DataSourceRef.Kind, claim.Spec.DataSourceRef.Name)
```
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Fixed.

File: pkg/controller/controller.go (hunk 13 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go:1122]: Fixed. https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1R1121

File: pkg/controller/controller.go (hunk 14 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: I fixed it as follows. Is this the fix you expected?
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1R1145-R1149

File: pkg/controller/controller.go (hunk 10 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Fixed.

File: pkg/controller/controller.go (hunk 11 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Fixed.

File: pkg/controller/controller.go (hunk 16 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Changed to check the featuregate in the dataSource function. Also, since it is confirmed at the beginning of the prepareProvision function, the confirmation of featuregate removed in the "IsGrant ", "getSnapshotSource ", and "getPVCSource " function.

File: pkg/controller/controller.go (hunk 12 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Fixed.

**Comment by ttakahashi21** [comment]
https://github.com/kubernetes-csi/external-provisioner/pull/805#discussion_r1028712172
I'm sorry. I was mistaken. No error occurs.

File: pkg/controller/controller.go (hunk 15 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Fixed.

File: pkg/controller/controller.go [hunk 19]
```diff
@@ -923,8 +940,16 @@ func (p *csiProvisioner) Provision(ctx context.Context, options controller.Provi
 	return pv, controller.ProvisioningFinished, nil
 }
 
-func (p *csiProvisioner) setCloneFinalizer(ctx context.Context, pvc *v1.PersistentVolumeClaim) error {
-	claim, err := p.claimLister.PersistentVolumeClaims(pvc.Namespace).Get(pvc.Spec.DataSource.Name)
+func (p *csiProvisioner) setCloneFinalizer(ctx context.Context, pvc *v1.PersistentVolumeClaim, dataSource *v1.ObjectReference) error {
+	var claim *v1.PersistentVolumeClaim
+	var err error
+
+	if len(dataSource.Namespace) > 0 {
```
Comment (msau42) [inline pkg/controller/controller.go:947]: Can `dataSource` also set the namespace field so we don't need this check?

File: pkg/controller/controller.go [hunk 20]
```diff
@@ -993,16 +1018,28 @@ func (p *csiProvisioner) getVolumeContentSource(ctx context.Context, claim *v1.P
...
 // returns the VolumeContentSource for the requested PVC
-func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentVolumeClaim, sc *storagev1.StorageClass) (*csi.VolumeContentSource, error) {
-	sourcePVC, err := p.claimLister.PersistentVolumeClaims(claim.Namespace).Get(claim.Spec.DataSource.Name)
+func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentVolumeClaim, sc *storagev1.StorageClass, dataSource *v1.ObjectReference) (*csi.VolumeContentSource, error) {
+
+	var sourcePVC *v1.PersistentVolumeClaim
+	var err error
+	if len(dataSource.Namespace) > 0 {
+		if claim.Namespace != dataSource.Namespace {
+			if ok, err := p.IsGranted(ctx, claim); err != nil || !ok {
```
Comment (msau42) [inline pkg/controller/controller.go:1027]: Can `IsGranted` also be moved to `dataSource`?

File: pkg/controller/controller.go (hunk 13 above)
Comment (msau42) [reply pkg/controller/controller.go:1122]: btw does the gateway controller have a similar method?

File: pkg/controller/controller_test.go [hunk 21]
```diff
@@ -4470,6 +4840,92 @@ func generatePVCForProvisionFromPVC(srcNamespace, srcName, scName string, reques
 	return provisionRequest
 }
 
+// generatePVCForProvisionFromPVC returns a ProvisionOptions with the requested settings
+func generatePVCForProvisionFromXnsPVC(srcNamespace, srcName, scName, kind string, refpvcNamespace, apiGroup *string, requestedBytes int64, volumeMode string, setnamespace bool) controller.ProvisionOptions {
+	deletePolicy := v1.PersistentVolumeReclaimDelete
+
+	var provisionRequest controller.ProvisionOptions
+	if setnamespace {
+		provisionRequest = controller.ProvisionOptions{
```
Comment (msau42) [inline pkg/controller/controller_test.go:4849]: It looks like almost everything is the same and the only difference is the namespace field. I think you can move this out of this `if` block and only change the namespace inside.

File: pkg/controller/controller_test.go [hunk 22]
```diff
@@ -4671,12 +5134,170 @@ func TestProvisionFromPVC(t *testing.T) {
...
+			refGrant:    newRefGrant("refGrant1", srcNamespace, ns1, coreapiGrp, "PersistentVolumeClaim"),
+		},
 	}
 
 	for k, tc := range testcases {
 		tc := tc
 		t.Run(k, func(t *testing.T) {
-			t.Parallel()
+			if !tc.xnsEnabled {
+				t.Parallel()
```
Comment (msau42) [inline pkg/controller/controller_test.go:5299]: Why does the xns feature require serial tests?

File: pkg/controller/controller_test.go [hunk 23]
```diff
@@ -4803,11 +5448,28 @@ func TestProvisionFromPVC(t *testing.T) {
...
+							},
+						}
+					} else if tc.volOpts.PVC.Spec.DataSource != nil {
+						volumeSource = csi.VolumeContentSource_Volume{
+							Volume: &csi.VolumeContentSource_VolumeSource{
+								VolumeId: tc.volOpts.PVC.Spec.DataSource.Name,
+							},
+						}
+					}
+				} else {
```
Comment (msau42) [inline pkg/controller/controller_test.go:5467]: Can this be collapsed into 2 conditions?

File: pkg/controller/controller.go [hunk 24]
```diff
@@ -1081,47 +1118,108 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
 	return volumeContentSource, nil
 }
 
+// TODO: This function should eventually be replaced by a common kubernetes library.
```
Comment (msau42) [inline pkg/controller/controller.go:1121]: For now this is a pretty important function from a security perspective. Can we move this out into a util file and have extensive unit tests for it?

File: pkg/controller/controller.go (hunk 16 above)
Comment (msau42) [reply pkg/controller/controller.go]: It looks much better now thanks!

File: pkg/controller/controller_test.go (hunk 21 above)
Comment (ttakahashi21) [reply pkg/controller/controller_test.go:4849]: Because the test did not work with feature gate enabled when it was executed in parallel test.

File: pkg/controller/controller_test.go (hunk 22 above)
Comment (ttakahashi21) [reply pkg/controller/controller_test.go:5299]: Because the test did not work with feature gate enabled when it was executed in parallel test.

File: pkg/controller/controller.go (hunk 19 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go:947]: It is possible.

File: pkg/controller/controller.go (hunk 20 above)
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Is it possible. Fixed code.

File: pkg/controller/controller.go (hunk 3 above)
Comment (robscott) [reply pkg/controller/controller.go]: I'm not sure I get the context here. Agree that ReferenceGrant is only useful/necessary to enable cross-namespace references. Same namespace references should not require a ReferenceGrant.

File: pkg/controller/controller_test.go (hunk 22 above)
Comment (msau42) [reply pkg/controller/controller_test.go:5299]: Hm that seems to imply there is some global shared state somewhere. I think it would be good to investigate more into that.

File: pkg/controller/controller_test.go (hunk 23 above)
Comment (ttakahashi21) [reply pkg/controller/controller_test.go]: Fixed.

File: pkg/controller/controller_test.go (hunk 21 above)
Comment (ttakahashi21) [reply pkg/controller/controller_test.go]: Fixed.

**Comment by ttakahashi21** [comment]
@msau42 
https://github.com/kubernetes-csi/external-provisioner/pull/805#discussion_r1036460110
I found two similar logics.

[cillum]
https://github.com/cilium/cilium/pull/21902/files#diff-c97054ccf2e8ead85e142ecda08c521b76ce13f924cc79dac145d380cd5bcf13R152-R174

[envoyproxy]
https://github.com/envoyproxy/gateway/pull/172/files#diff-be48a4416dad5f211bed7f15e1cc13e8cdfaa1378a7d5be349d8bb06e4b2054eR706-R745

File: pkg/controller/controller_test.go (hunk 22 above)
Comment (ttakahashi21) [reply pkg/controller/controller_test.go:5265]: `DefaultFeatureGate` defined in [utilfeature](https://github.com/kubernetes/apiserver/blob/master/pkg/util/feature/feature_gate.go#L32) seems a global value, therefore setting [value](https://github.com/kubernetes-csi/external-provisioner/blob/3bfb91586acdcce7c8cd1796ebe0ab2770cf880d/vendor/k8s.io/component-base/featuregate/testing/feature_gate.go#L35) during tests causes changes to a global value in each test.

File: pkg/controller/util_test.go [hunk 25]
```diff
@@ -0,0 +1,175 @@
...
+	fromNamespace := "ns1"
+	fromsrcName := "test-dataSource"
+	toNamespace := "ns2"
+
+	type testcase struct {
+		expectErr    bool
+		volOpts      controller.ProvisionOptions
+		refGrantList []*gatewayv1beta1.ReferenceGrant
+	}
+	testcases := map[string]testcase{
```
Comment (msau42) [inline pkg/controller/util_test.go:64]: Can you add a test case for nil datasource apigroup?

File: pkg/controller/util_test.go [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+		"Allowed to access dataSource for xns PVC with refGrant": {
+			volOpts:      generatePVCForProvisionFromXnsdataSource(toNamespace, fromsrcName, "", pvcapiKind, &fromNamespace, &coreapiGrp, requestedBytes, "", true),
+			refGrantList: newRefGrantList("refGrant1", fromNamespace, coreapiGrp, pvcapiKind, toNamespace, coreapiGrp, pvcapiKind, "", false),
```
Comment (msau42) [inline pkg/controller/util_test.go:67]: I know we do this in our other test files, but I think it's a little difficult to read the test case with this style of constructor because I need to remember the ordering of the arguments and which field they correspond to. It's perhaps more lines, but I think it's more readable to directly reference the object in the test case definition.

If you want to reduce the boilerplate, maybe you can still have a constructor, but have it take a `ReferenceGrantFrom` and a `ReferenceGrantTo` as arguments?

Similar comment for `generatePVC...` Keep the constructor to reduce on the common boilerplate, but have it take `DataSourceRef` type as an input, which is what changes across each test case.

**Review by msau42 — COMMENTED** [review]
Thanks for all the great work! I just have a few unit test suggestions.

File: pkg/controller/controller_test.go (hunk 22 above)
Comment (ttakahashi21) [reply pkg/controller/controller_test.go:6265]: I added the link of the issue I created as a comment

File: pkg/controller/controller.go [hunk 27]
```diff
@@ -923,8 +940,12 @@ func (p *csiProvisioner) Provision(ctx context.Context, options controller.Provi
 	return pv, controller.ProvisioningFinished, nil
 }
 
-func (p *csiProvisioner) setCloneFinalizer(ctx context.Context, pvc *v1.PersistentVolumeClaim) error {
-	claim, err := p.claimLister.PersistentVolumeClaims(pvc.Namespace).Get(pvc.Spec.DataSource.Name)
+func (p *csiProvisioner) setCloneFinalizer(ctx context.Context, pvc *v1.PersistentVolumeClaim, dataSource *v1.ObjectReference) error {
+	var claim *v1.PersistentVolumeClaim
```
[the remark this answers is not in the document]
Comment (ttakahashi21) [reply pkg/controller/controller.go]: Fixed.

File: pkg/controller/util_test.go (hunk 25 above)
Comment (ttakahashi21) [reply pkg/controller/util_test.go]: https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-640957054096359bf678ef9f2789673a60814fba8286165ad298f5eba881536aR215-R222
I added.

File: pkg/controller/util_test.go (hunk 26 above)
Comment (ttakahashi21) [reply pkg/controller/util_test.go]: util_test.go created a base test case. And I modified the base test case and modified it to add the test case.
Along with the above modifications, TestProvisionFromSnapshot and TestProvisionFromPVC have also been modified.

In util.go I added the following just in case:
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-ded124c02add3467ce103641d0cd28436ea69cf37c5906470b3d519009d950b7R16-R20

As it turned out to be necessary in the test case below.
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-640957054096359bf678ef9f2789673a60814fba8286165ad298f5eba881536aR166-R170

This condition is unnecessary in the controller because it is acquired by specifying the namespace, but I added it just in case.
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1R1936-R1937

**Comment by msau42** [comment]
/lgtm
/approve

Thanks for adding all the comprehensive unit tests!

For the release note, can you update it to say "Add support for cross-namespace data sources alpha feature"
